fix(cli): error on missing config file

sgaist opened the pull request:
## Description

This PR will make trivy error if the configuration file passed in parameter cannot be found.

However, to stay consistent with the current behaviour

## Related issues
- Close #7028

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [ ] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

simar7 commented:
Thank you for the PR. Couple of comments:

1. The level should be Warning. The user should be notified that the results aren't as desired given their input isn't right/missing.
2. This should be extended to other config files that trivy uses, for instance the `.trivyignore.yaml`. Ultimately, any specified user input, if not correct, should be flagged by Trivy so users know what to expect.

sgaist commented:
@simar7 

1) I'll update the code suggestion and submit it

2) I agree. I'd suggest making a dedicated task list to ensure we don't miss such an option

knqyf263 commented:
@simar7 It currently returns an error when a user specifies `--config` and it is not found. It should be fine.

```
$ trivy image --config trivy.yaml alpine:3.20
2024-07-16T11:49:31+04:00       FATAL   Fatal error     config file "trivy.yaml" loading error: open trivy.yaml: no such file or directory
```

knqyf263 commented on pkg/commands/app.go:
```diff
@@ -151,14 +151,16 @@ func loadPluginCommands() []*cobra.Command {
-			log.Debug("Config file not found", log.FilePath(configFile))
-			return nil
+			if !pathChanged {
+				log.Debug("Config file not found", log.FilePath(configFile))
```
I'd keep it at the debug level. Otherwise, it shows a message every time, which is annoying for most users.

BTW, this log is called before initializing the logger. So, it displays nothing now. We also need to fix it.
https://github.com/aquasecurity/trivy/blob/c8a7abd3b508975fcf10c254d13d1a2cd42da657/pkg/commands/app.go#L214

sgaist replied:
@knqyf263 Understood, I'll put it back as debug.

As for the logger initialization, don't we have a chicken and egg problem here ? Or should it be initialized with defaults and then re-initialized based on the configuration ?

knqyf263 replied:
>As for the logger initialization, don't we have a chicken and egg problem here ? 

Yes.

>Or should it be initialized with defaults and then re-initialized based on the configuration ?

I think you're right. We should initialize the logger twice.

knqyf263 commented on .gitignore:
```diff
@@ -39,3 +39,6 @@ dist
 cmd/trivy/trivy
+
+# bin folder
+bin/
```
Since the Trivy project doesn't have `bin/`, I'd suggest excluding it on your end.
https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files#excluding-local-files-without-creating-a-gitignore-file
https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files#configuring-ignored-files-for-all-repositories-on-your-computer

sgaist replied:
While Trivy itself doesn't have that directory, it's was created to contain `golangci-lint` when I ran the `mage` lint related commands hence I added it to `.gitignore`.
Isn't that a good reason ?
That said, thanks for the excellent pointers, I learned something new.

knqyf263 replied:
I see. It's supposed to be installed under `$GOPATH/bin`.
https://github.com/aquasecurity/trivy/blob/b76a7250912cfc028cfef743f0f98cd81b39f8aa/magefiles/magefile.go#L25-L26

I guess you don't have `$GOPATH`, then `$PWD/bin` was created. It's sort of a bug in our Magefile, and not intended. In your environment, does `go env GOPATH` return something?

sgaist replied:
So something is a bit strange here: `echo $GOPATH` returns empty but if I call `go env GOPATH` then it points to `/Users/username/go` however there's no `bin` folder in there.

knqyf263 replied:
If I remember correctly, the default GOPATH is baked in Go so that Go can work without the `$GOPATH` env. We should fix the magefile. So, we don't need to add `bin/` in .gitignore since it's not intended.

sgaist replied:
Fair enough, I have dropped the related commit.

knqyf263 commented on pkg/commands/app.go:
```diff
@@ -151,14 +151,16 @@ func loadPluginCommands() []*cobra.Command {
-			log.Debug("Config file not found", log.FilePath(configFile))
-			return nil
+			if !pathChanged {
+				log.Warnf("Default config file %q not found, using built in values", log.FilePath(configFile))
```
Please switch it back to debug.

sgaist replied:
Done

knqyf263 reviewed (commented):
We can fix the logger initialization problem in another PR. This change is already valuable. We should merge this PR.

knqyf263 commented on pkg/commands/app.go:
```diff
@@ -195,13 +197,15 @@ func NewRootCommand(globalFlags *flag.GlobalFlagGroup) *cobra.Command {
 				return xerrors.Errorf("flag bind error: %w", err)
 			}
 
+			log.InitLogger(true, false)
```
I'm sorry, but this is not what I meant. I was supposed to describe the details after this PR. Now, even if `--debug` is not specified, Trivy always shows debug messages, which confuses users.

```
$ trivy image alpine:3.20
2024-07-18T12:56:11+04:00       DEBUG   Default config file "file_path=trivy.yaml" not found, using built in values
2024-07-18T12:56:11+04:00       DEBUG   Cache dir       dir="/Users/teppei/Library/Caches/trivy"
```

We can fix it in another PR. Could you revert?

sgaist replied:
Sorry, it was a test that I didn't meant to put in this PR.

knqyf263 reviewed (approved):
LGTM. Thanks!

sgaist commented:
@knqyf263 You're welcome !

I just fixed the test issue: it's the OS specific part of the string that makes it fail.

I removed that part from the string check however would you rather have a fixed message ?

knqyf263 commented:
>I removed that part from the string check however would you rather have a fixed message ?

It looks good to me.

sgaist commented:
I missed that the tests have failed and I can't see the content anymore, is there something I can do trigger them again beside rebasing my branch ?

knqyf263 commented:
You can merge the main branch and let's see how it goes.

sgaist commented:
Done as requested
